File: solution/3425/고스택.py
f=open('고스택.txt', 'r')
input=f.readline

# from sys import stdin
# input=stdin.readline

cmddict={'NUM':0, 'POP':1, 'INV':2, 'DUP':3, 'SWP':4, 'ADD':5, 'SUB':6, 'MUL':7, 'DIV':8, 'MOD':9}

inp=input().rstrip()
while inp[0]!='Q':
    cmds=[]
    while inp[0]!='E':
        if len(inp)>3:
            cmd, num = inp.split()
            cmd, num = cmddict[cmd], int(num)
        else: cmd, num = cmddict[inp], None
        cmds.append([cmd, num])
        inp=input().rstrip()
    n=int(input())
    for _ in range(n):
        stack=[int(input())]
        for cmd, num in cmds:
            try:
                if cmd==1: stack.pop()
                elif cmd==2: stack[-1]*=-1
                elif cmd==3: stack.append(stack[-1])
                elif cmd==4: stack[-1], stack[-2] = stack[-2], stack[-1]
                elif cmd>4:
                    a, b = stack.pop(), stack.pop()
                    if cmd==5: stack.append(a+b)
                    elif cmd==6: stack.append(b-a)
                    elif cmd==7: stack.append(a*b)
                    elif cmd==8:
                        if (a<0 and b<0) or (a>0 and b>0): stack.append(abs(b)//abs(a))
                        else: stack.append(-(abs(b)//abs(a)))
                    elif cmd==9:
                        if b<0: stack.append(-abs(b%a))
                        else: stack.append(abs(b%a))
                elif cmd==0: stack.append(num)
            except:
                print('ERROR')
                break
        else:
            if not stack: print('ERROR')
            elif len(stack)>1 or abs(stack[-1])>1000000000: print('ERROR')
            else: print(stack[-1])
    inp=input().rstrip()
    if not inp: inp=input().rstrip()
    if inp[0]!='Q':print()

File: solution/3425/고스택.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;

    static Stack<Integer> stack;
    static ArrayList<String> Task;

    public static void main(String[] args) throws Exception {
//        br = new BufferedReader(new InputStreamReader(System.in));
        br = new BufferedReader(new InputStreamReader(new FileInputStream("./input.txt")));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));

        while (true){
            String input = br.readLine();
            if (input.equals("QUIT"))
                break;

            Task = new ArrayList<>();
            while (!input.equals("END")){
                Task.add(input);
                input = br.readLine();
            }

            int N = Integer.parseInt(br.readLine());
            for (int i = 0; i < N; i++) {
                stack = new Stack<>();
                stack.push(Integer.parseInt(br.readLine()));
                if (run())
                    bw.write(stack.pop()+"\n");
                else
                    bw.write("ERROR\n");
            }
            bw.write("\n");
            br.readLine();
        }

        bw.flush();
        bw.close();
        br.close();
    }
    static boolean run(){
        for (String line : Task) {
            st = new StringTokenizer(line);
            String cmd = st.nextToken();

            if (cmd.equals("NUM")){
                stack.push(Integer.parseInt(st.nextToken()));
            }
            else if (cmd.equals("POP")){
                if (stack.empty())
                    return false;
                stack.pop();
            }
            else if (cmd.equals("INV")){
                if (stack.empty())
                    return false;
                stack.push(-stack.pop());
            }
            else if (cmd.equals("DUP")){
                if (stack.empty())
                    return false;
                stack.push(stack.peek());
            }
            else if (cmd.equals("SWP")){
                if (stack.empty())
                    return false;
                int A = stack.pop();
                if (stack.empty())
                    return false;
                int B = stack.pop();
                stack.push(A);
                stack.push(B);
            }
            else if (cmd.equals("ADD")){
                if (stack.empty())
                    return false;
                int A = stack.pop();
                if (stack.empty())
                    return false;
                int B = stack.pop();
                if (Math.abs(A+B)>1000000000)
                    return false;
                stack.push(A+B);
            }
            else if (cmd.equals("SUB")){
                if (stack.empty())
                    return false;
                int A = stack.pop();
                if (stack.empty())
                    return false;
                int B = stack.pop();
                if (Math.abs(B-A)>1000000000)
                    return false;
                stack.push(B-A);
            }
            else if (cmd.equals("MUL")){
                if (stack.empty())
                    return false;
                long A = stack.pop();
                if (stack.empty())
                    return false;
                long B = stack.pop();
                if (Math.abs(A*B)>1000000000)
                    return false;
                stack.push((int) (A*B));
            }
            else if (cmd.equals("DIV")){
                if (stack.empty())
                    return false;
                int A = stack.pop();
                if (stack.empty() || A==0)
                    return false;
                int B = stack.pop();
                if (((A>0 && B<0) || (A<0 && B>0)))
                    stack.push(-(Math.abs(B)/Math.abs(A)));
                else
                    stack.push(Math.abs(B)/Math.abs(A));
            }
            else if (cmd.equals("MOD")){
                if (stack.empty())
                    return false;
                int A = stack.pop();
                if (stack.empty() || A==0)
                    return false;
                int B = stack.pop();
                if (B<0)
                    stack.push(-(Math.abs(B)%Math.abs(A)));
                else
                    stack.push(Math.abs(B)%Math.abs(A));
            }
        }
        if (stack.size()!=1)
            return false;
        return true;
    }
}